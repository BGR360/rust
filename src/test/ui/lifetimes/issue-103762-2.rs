// Failure cases for #103762.

trait Trait {}

// Case 1: '_ in trait object bound without any other anonymous/elided lifetimes.
// Pre-1.64 rejected this, but 1.64 did not. Make sure it does.
fn a1(_: Box<dyn Trait + '_>) -> &str { loop {} } //~ ERROR missing lifetime specifier
fn a2(_: Box<Trait + '_>) -> &str { loop {} } //~ ERROR missing lifetime specifier
//~^ WARNING deprecated

// Case 2: Inside fn() type.
// Pre-1.64 and 1.64 rejected this, should still reject.
fn b1(_: fn(&(dyn Trait + '_)) -> &str) { loop {} } //~ ERROR missing lifetime specifier
fn b2(_: fn(&(Trait + '_)) -> &str) { loop {} } //~ ERROR missing lifetime specifier
//~^ WARNING deprecated

// Case 3: Inside generic bound.
// Pre-1.64 and 1.64 rejected this, should still reject.
fn c1<T: Trait + '_>(_: &T) -> &str { loop {} } //~ ERROR `'_` cannot be used here
fn c2<T>(_: &T) -> &str where T: Trait + '_ { loop {} } //~ ERROR `'_` cannot be used here

// Case 4.a: Inside impl Trait without any other anonymous/elided lifetimes.
//  * Pre-1.64 rejected this because it thought `+ '_` needed to be a named lifetime,
//    and did not report on `&str`.
//  * 1.64 rejected this because of unstable anonymous_lifetime_in_impl_trait,
//    and because `&str` needs a named lifetime.
// It makes sense to keep doing what 1.64 did.
fn d1(_: impl Trait + '_) //~ ERROR anonymous lifetimes in `impl Trait` are unstable
    -> &str { loop {} } //~ ERROR missing lifetime specifier

// Case 4.b: Inside impl Trait with other anonymous/elided lifetimes.
//  * Pre-1.64 rejected these because it thought `+ '_` needed to be a named lifetime.
//  * 1.64 rejected these because of unstable anonymous_lifetime_in_impl_trait,
//    and because it thought `&str` needed a named lifetime parameter.
// It should continue reporting on the unstable feature, but not on `&str`, since we
// shouldn't count `+ '_` as a new elision candidate.
fn d2(_: &(impl Trait + '_)) //~ ERROR anonymous lifetimes in `impl Trait` are unstable
    -> &str { loop {} }
fn d3(_: &i32, _: impl Trait + '_) //~ ERROR anonymous lifetimes in `impl Trait` are unstable
    -> &str { loop {} }
fn d4(_: impl Trait + '_, _: &i32) //~ ERROR anonymous lifetimes in `impl Trait` are unstable
    -> &str { loop {} }

fn main() {}
