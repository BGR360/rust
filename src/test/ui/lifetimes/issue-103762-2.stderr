error[E0106]: missing lifetime specifier
  --> $DIR/issue-103762-2.rs:7:34
   |
LL | fn a1(_: Box<dyn Trait + '_>) -> &str { loop {} }
   |                                  ^ expected named lifetime parameter
   |
   = help: this function's return type contains a borrowed value, but there is no value for it to be borrowed from
help: consider using the `'static` lifetime
   |
LL | fn a1(_: Box<dyn Trait + '_>) -> &'static str { loop {} }
   |                                   +++++++

error[E0106]: missing lifetime specifier
  --> $DIR/issue-103762-2.rs:8:30
   |
LL | fn a2(_: Box<Trait + '_>) -> &str { loop {} }
   |                              ^ expected named lifetime parameter
   |
   = help: this function's return type contains a borrowed value, but there is no value for it to be borrowed from
help: consider using the `'static` lifetime
   |
LL | fn a2(_: Box<Trait + '_>) -> &'static str { loop {} }
   |                               +++++++

error[E0637]: `'_` cannot be used here
  --> $DIR/issue-103762-2.rs:19:18
   |
LL | fn c1<T: Trait + '_>(_: &T) -> &str { loop {} }
   |                  ^^ `'_` is a reserved lifetime name

error[E0637]: `'_` cannot be used here
  --> $DIR/issue-103762-2.rs:20:42
   |
LL | fn c2<T>(_: &T) -> &str where T: Trait + '_ { loop {} }
   |                                          ^^ `'_` is a reserved lifetime name

error[E0106]: missing lifetime specifier
  --> $DIR/issue-103762-2.rs:29:8
   |
LL |     -> &str { loop {} }
   |        ^ expected named lifetime parameter
   |
   = help: this function's return type contains a borrowed value, but there is no value for it to be borrowed from
help: consider using the `'static` lifetime
   |
LL |     -> &'static str { loop {} }
   |         +++++++

error[E0228]: the lifetime bound for this object type cannot be deduced from context; please supply an explicit bound
  --> $DIR/issue-103762-2.rs:7:14
   |
LL | fn a1(_: Box<dyn Trait + '_>) -> &str { loop {} }
   |              ^^^^^^^^^^^^^^

warning: trait objects without an explicit `dyn` are deprecated
  --> $DIR/issue-103762-2.rs:8:14
   |
LL | fn a2(_: Box<Trait + '_>) -> &str { loop {} }
   |              ^^^^^^^^^^
   |
   = warning: this is accepted in the current edition (Rust 2015) but is a hard error in Rust 2021!
   = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2021/warnings-promoted-to-error.html>
   = note: `#[warn(bare_trait_objects)]` on by default
help: use `dyn`
   |
LL | fn a2(_: Box<dyn Trait + '_>) -> &str { loop {} }
   |              +++

error[E0228]: the lifetime bound for this object type cannot be deduced from context; please supply an explicit bound
  --> $DIR/issue-103762-2.rs:8:14
   |
LL | fn a2(_: Box<Trait + '_>) -> &str { loop {} }
   |              ^^^^^^^^^^

warning: trait objects without an explicit `dyn` are deprecated
  --> $DIR/issue-103762-2.rs:14:15
   |
LL | fn b2(_: fn(&(Trait + '_)) -> &str) { loop {} }
   |               ^^^^^^^^^^
   |
   = warning: this is accepted in the current edition (Rust 2015) but is a hard error in Rust 2021!
   = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2021/warnings-promoted-to-error.html>
help: use `dyn`
   |
LL | fn b2(_: fn(&(dyn Trait + '_)) -> &str) { loop {} }
   |               +++

error[E0658]: anonymous lifetimes in `impl Trait` are unstable
  --> $DIR/issue-103762-2.rs:28:23
   |
LL | fn d1(_: impl Trait + '_)
   |                       ^^ expected named lifetime parameter
   |
   = help: add `#![feature(anonymous_lifetime_in_impl_trait)]` to the crate attributes to enable
help: consider introducing a named lifetime parameter
   |
LL | fn d1<'a>(_: impl Trait + '_'a )
   |      ++++                   ++

error[E0658]: anonymous lifetimes in `impl Trait` are unstable
  --> $DIR/issue-103762-2.rs:37:25
   |
LL | fn d2(_: &(impl Trait + '_))
   |                         ^^ expected named lifetime parameter
   |
   = help: add `#![feature(anonymous_lifetime_in_impl_trait)]` to the crate attributes to enable
help: consider introducing a named lifetime parameter
   |
LL | fn d2<'a>(_: &(impl Trait + '_'a ))
   |      ++++                     ++

error[E0658]: anonymous lifetimes in `impl Trait` are unstable
  --> $DIR/issue-103762-2.rs:39:32
   |
LL | fn d3(_: &i32, _: impl Trait + '_)
   |                                ^^ expected named lifetime parameter
   |
   = help: add `#![feature(anonymous_lifetime_in_impl_trait)]` to the crate attributes to enable
help: consider introducing a named lifetime parameter
   |
LL | fn d3<'a>(_: &i32, _: impl Trait + '_'a )
   |      ++++                            ++

error[E0658]: anonymous lifetimes in `impl Trait` are unstable
  --> $DIR/issue-103762-2.rs:41:23
   |
LL | fn d4(_: impl Trait + '_, _: &i32)
   |                       ^^ expected named lifetime parameter
   |
   = help: add `#![feature(anonymous_lifetime_in_impl_trait)]` to the crate attributes to enable
help: consider introducing a named lifetime parameter
   |
LL | fn d4<'a>(_: impl Trait + '_'a , _: &i32)
   |      ++++                   ++

error: aborting due to 11 previous errors; 2 warnings emitted

Some errors have detailed explanations: E0106, E0228, E0637, E0658.
For more information about an error, try `rustc --explain E0106`.
